src/main/java/org/humancellatlas/ingest/state/persistence/AutoLoader.java

```diff
@@ -29,23 +42,113 @@
+        if (envelope.isEmpty()) {
+            deleteStateMachine(stateMachine);
+            return Optional.empty();
+        } else {
+            return Optional.of(updateState(stateMachine, envelope.get()));
+        }
+    }
+
+    private StateMachine<SubmissionState, SubmissionEvent> updateState(StateMachine<SubmissionState, SubmissionEvent> stateMachine, SubmissionEnvelopeReference envelope) throws UnrecognisedSubmissionStateException {
+        SubmissionState currentState = stateMachine.getState().getId();
+        SubmissionState correctState = convertState(envelope.getState());
+
+        // Note that some "ongoing" states like Draft/Validating/Invalid/Processing/Exporting can have extended states which
+        // tracks some metadata in a submission. It would be more complicated to sync these extended states from core.
+        // Syncing the state to a finished state like Valid, Submitted, Archived, Exported, Cleanup, Complete should be safe.
+        // It might be best to redesign how we use the state machines and not have a separate state tracker.
+
+        List<SubmissionState> finishedStates = Arrays.asList(VALID, SUBMITTED, ARCHIVED, EXPORTED, CLEANUP, COMPLETE);
+
+        if (!currentState.equals(correctState) && finishedStates.contains(correctState)) {
+            stateMachine.getStateMachineAccessor().doWithAllRegions(
+                    access -> access.resetStateMachine(
+                            new DefaultStateMachineContext<>(correctState, null, null, null, null, stateMachine.getId())
+                    )
+            );
+
+            String envelopeUuid = envelope.getUuid();
+            log.info(String.format("Synced %s from state %s to state: %s, with id: %s",
+                    envelopeUuid, currentState, correctState, stateMachine.getId()));
+        }
+        return stateMachine;
+    }
+
+    private void deleteStateMachine(StateMachine<SubmissionState, SubmissionEvent> stateMachine) {
+        Optional<RedisRepositoryStateMachine> redisStateMachine = stateMachineRepository.findById(stateMachine.getId());
+        stateMachineRepository.delete(redisStateMachine.get());
+        log.info(String.format("Deleted statemachine, id: %s, state:%s", stateMachine.getId(), stateMachine.getState().toString()));
+    }
+
+    private SubmissionState convertState(String state) {
+        try {
+            SubmissionState correctState = SubmissionState.fromString(state);
+            return correctState;
+        } catch (UnrecognisedSubmissionStateException e) {
+            String desc = String.format("A submission state value is not recognised in the state tracker: %s", e.getMessage());
+            String action = "Consider defining a new SubmissionState enum value";
+            handleException(desc, action, e);
+        }
+        return null;
```

Comment:
Is this line reachable?

Reply:
No, but it gives me a compilation error that it's missing a return statement